add hydrogen deploy error handling & oxygen api service

tempor1s opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [Feature] (if applicable)
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Use a draft PR while it’s a work in progress
-->

### WHY are these changes introduced?

Closes https://github.com/Shopify/oxygen-platform/issues/528

This PR introduces error handling and retry logic for the various calls that we make to DMS so that we can be more resilient / provide context to the user on why their deploy might be failing, etc. I also greatly increased the scope of what the unit tests cover.

I also created a new service within the CLI kit that is used for interacting with DMS as it seems this is the pattern and convention that is being used to interact with other APIs (identity, partners, shopify, etc) and removes the need for the CLI kit to export internal functionality (namely `buildHeaders`)

This PR also places the zip file of built assets into a temporary directory to follow the convention that `oxygenctl` is currently using instead of just placing the zip in the dist.

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes

lynchv commented on packages/cli-hydrogen/src/cli/services/deploy.test.ts:
```diff
@@ -93,6 +93,54 @@ describe('deployToOxygen()', () => {
     expect(mockedHealthCheck).toHaveBeenCalledTimes(3)
   })
 
+  it('does retry for createDeployment if the error is recoverable', async () => {
```
Could throw these in describe blocks
```suggestion
describe('when error is recoverable')
  it('retry createDeployment')
  
  it('retry uploadDeployment')
```

lynchv commented on packages/cli-hydrogen/src/cli/services/deploy/upload.ts:
```diff
@@ -8,14 +8,17 @@ import {UploadDeploymentQuery} from './graphql/upload_deployment.js'
-    headers,
-    url: getOxygenAddress(config.oxygenAddress),
-  })
+export const TooManyRequestsError = () => {
```
Should put those in `packages/cli-hydrogen/src/cli/services/deploy/error.ts`

lynchv commented on packages/cli-hydrogen/src/cli/services/deploy/upload.ts:
```diff
@@ -26,36 +29,73 @@ export const createDeployment = async (config: ReqDeployConfig): Promise<CreateD
 
 export const uploadDeployment = async (config: ReqDeployConfig, deploymentID: string): Promise<string> => {
-  let headers = await api.buildHeaders(config.deploymentToken)
+  let deploymentData: UploadDeploymentResponse | undefined
```
I don't think this is needed, could simply be defined as a `const` below.

tempor1s commented:
## Force Push Notes
- Extracted errors into separate file
- Changed `unrecoverable` errors to throw instead of just returning to stop the tasks from continuing.
- Extracted retry logic into separate method
- Improve tests to cover new exception functionality

lynchv commented on packages/cli-hydrogen/src/cli/services/deploy.ts:
```diff
@@ -95,3 +99,24 @@ export async function deployToOxygen(_config: DeployConfig) {
+async function shouldRetryOxygenCall(
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  task: ui.ListrTaskWrapper<TaskContext, any>,
+  backoffPolicy: number[],
```
I think we could make the `backoffPolicy` a global var and not pass it down?

lynchv reviewed (approved):
lgtm

tempor1s commented:
Hey @amcaplan / @pepicrft would you mind taking a look at this PR before I merge it since I am adding a new service to the CLI kit? Thanks!

tempor1s commented on packages/cli-hydrogen/src/cli/services/deploy.test.ts:
```diff
@@ -77,20 +77,75 @@ describe('deployToOxygen()', () => {
+  })
+
+  describe('when error is unrecoverable', async () => {
+    it('does retry for createDeployment', async () => {
```
Good catch!
